### PR TITLE
feat(api): Add precision types support

### PR DESCRIPTION
**Pull request by w4bremer**
Add support for float64/double and int64. This only works with UE5. For UE4 the solution file has to switch to a different API module.

### CODE REVIEW

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Tests/TbSimpleSimpleArrayInterfaceTests.cpp [hunk 1]
```diff
@@ -40,6 +40,26 @@ bool UTbSimpleSimpleArrayInterfaceImplementationPropertyPropIntTest::RunTest(con
 	return true;
 }
 
+IMPLEMENT_SIMPLE_AUTOMATION_TEST(UTbSimpleSimpleArrayInterfaceImplementationPropertyPropInt32Test, "TbSimple.SimpleArrayInterface.Implementation.Property.PropInt32", EAutomationTestFlags::ApplicationContextMask | EAutomationTestFlags::SmokeFilter)
+bool UTbSimpleSimpleArrayInterfaceImplementationPropertyPropInt32Test::RunTest(const FString& Parameters)
+{
+	// Do implement test here
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Tests/TbSimpleSimpleArrayInterfaceTests.cpp:46]: Maybe  "// Implement test here" instead ? Or what's the exact purpose of the comment?

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Implementation/TbSimpleSimpleInterface.cpp [hunk 2]
```diff
@@ -184,13 +320,41 @@ int32 UTbSimpleSimpleInterface::FuncInt_Implementation(int32 ParamInt)
 	return 0;
 }
 
+int32 UTbSimpleSimpleInterface::FuncInt32_Implementation(int32 ParamInt32)
+{
+	(void)ParamInt32;
+	// do business logic here
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Implementation/TbSimpleSimpleInterface.cpp:326]: Use of lower case and capital beginning letters in comments

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Public/Implementation/TbSimpleSimpleInterface.h [hunk 3]
```diff
@@ -92,22 +140,42 @@ class TBSIMPLE_API UTbSimpleSimpleInterface : public UObject, public ITbSimpleSi
...
 	void BroadcastPropFloatChanged_Implementation(float PropFloat) override;
 
+	void BroadcastPropFloat32Changed_Implementation(float PropFloat32) override;
+
+	void BroadcastPropFloat64Changed_Implementation(double PropFloat64) override;
+
 	void BroadcastPropStringChanged_Implementation(const FString& PropString) override;
 
 private:
 	// properties - local copy - use setter functions to emit changed signals
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Public/Implementation/TbSimpleSimpleInterface.h:172]: Lowercase vs capital beginning letters for comments

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp [hunk 4]
```diff
@@ -372,11 +620,31 @@ void UTbSimpleSimpleInterfaceOLinkClient::emitSignal(const std::string& signalNa
 		Execute_BroadcastSigInt(this, args[0].get<int32>());
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp:620]: What is args here? args[0] could probably be defined somewhere and use throughout code for clarity. Counting 14 usages.

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp [hunk 5]
```diff
@@ -316,6 +532,22 @@ void UTbSimpleSimpleArrayInterfaceOLinkClient::applyState(const nlohmann::json&
 			Execute_BroadcastPropIntChanged(this, PropInt);
 		}
 	}
+	if (fields.contains("propInt32"))
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp:535]: Avoid nested ifs here for brevity and clarity?

if ( fields.contains("propInt32") && PropInt32 != fields["propInt32"].get<TArray<int32>>() )

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp [hunk 6]
```diff
@@ -279,6 +457,44 @@ TArray<float> UTbSimpleSimpleArrayInterfaceOLinkClient::FuncFloat_Implementation
 	return Promise.GetFuture().Get();
 }
 
+TArray<float> UTbSimpleSimpleArrayInterfaceOLinkClient::FuncFloat32_Implementation(const TArray<float>& ParamFloat32)
+{
+	if (!m_sink->IsReady())
+	{
+		UE_LOG(LogTemp, Warning, TEXT("%s has no node"), UTF8_TO_TCHAR(m_sink->olinkObjectName().c_str()));
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp:464]: Is there an ability to turn logging off?

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/Monitor/TbSimple.trace.cpp [hunk 7]
```diff
@@ -188,13 +328,41 @@ void TbSimpleSimpleArrayInterfaceTracer::trace_callFuncInt(const TArray<int32>&
 	Tracer::instance()->call("tb.simple/SimpleArrayInterface#funcInt", fields_);
 }
 
+void TbSimpleSimpleArrayInterfaceTracer::trace_callFuncInt32(const TArray<int32>& ParamInt32)
+{
+	nlohmann::json fields_;
```
Comment (zrrbite) [inline goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/Monitor/TbSimple.trace.cpp:333]: Some inconsistencies in local scope variable naming, using trailing underscores  (e.g. fields_ vs most often not)

**Review by zrrbite — CHANGES_REQUESTED** [review]
Does this commit need apigear/testbed.simpleUE5.module.yaml? I just read in the documentation that it's fully compatible with UE5.

Would the code generally benefit from more comments? So that it is easier to make the connection between IDL and generated code, etc.

File: apigear/testbed.simple.module.yaml [hunk 8]
```diff
@@ -16,30 +16,46 @@ interfaces:
...
           - { name: paramBool, type: bool }
-        return:
-          type: bool
+        return: { type: bool }
       - name: funcInt
         params:
           - { name: paramInt, type: int }
-        return:
-          type: int
+        return: { type: int }
```
[the remark this answers is not in the document]
Comment (w4bremer) [reply apigear/testbed.simple.module.yaml:34]: True. I just took it from another template repository which had the different format.
However, I checked that the content is the same aside from the format.
We plan to align the testbed APIs for all templates in the near future.

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp (hunk 6 above)
Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp:464]: Not yet, created issue for this https://github.com/apigear-io/template-unreal/issues/33

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/Monitor/TbSimple.trace.cpp (hunk 7 above)
Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/Monitor/TbSimple.trace.cpp:333]: Created issue for this https://github.com/apigear-io/template-unreal/issues/34

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp (hunk 5 above)
Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleArrayInterfaceOLinkClient.cpp:535]: Created issue for this https://github.com/apigear-io/template-unreal/issues/35

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp (hunk 4 above)
Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp:620]: args is a json array which on the first entry contains the signal data to be sent.
We could store it a temporary variable before broadcasting it. Would this be better?

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Tests/TbSimpleSimpleArrayInterfaceTests.cpp (hunk 1 above)
Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Tests/TbSimpleSimpleArrayInterfaceTests.cpp:46]: It is meant for the user of the generated code to add the test logic in this function.
Don't think we need the "Do" here, but would not start a new PR/commit just to remove it now. Ok?

File: goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp (hunk 4 above)
Comment (zrrbite) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp:620]: I think that would be good, yes.

Comment (w4bremer) [reply goldenmaster/Plugins/TbSimple/Source/TbSimple/Private/Generated/OLink/TbSimpleSimpleInterfaceOLinkClient.cpp:620]: Create issue for this https://github.com/apigear-io/template-unreal/issues/36

**Comment by w4bremer** [comment]
> Does this commit need apigear/testbed.simpleUE5.module.yaml? I just read in the documentation that it's fully compatible with UE5.
No, the current `apigear/testbed.simple.module.yaml` shall be the new default going forward. Only if we want to assure UE4 comability we should use the `apigear/testbed.simpleUE4.module.yaml` module.

> Would the code generally benefit from more comments? So that it is easier to make the connection between IDL and generated code, etc.
Yes, but there is also the folder structure. Generated code, which will be automatically overwritten, should go to `generated` folder. The same goes with real implementation and the `implementation` folder.
More will be explained in the documentation and code comments

**Review by zrrbite — APPROVED** [review]

**Pull request merged**